baseplate/metrics.py

```diff
@@ -255,18 +279,71 @@ def increment(self, delta=1, sample_rate=1.0):
         :param float delta: The amount to increase the counter by.
         :param float sample_rate: What rate this counter is sampled at. [0-1].
 
+        """
+        self.send(delta, sample_rate)
+
+    def decrement(self, delta=1, sample_rate=1.0):
+        """Decrement the counter.
+
+        This is equivalent to :py:meth:`increment` with delta negated.
+
+        """
+        self.increment(delta=-delta, sample_rate=sample_rate)
+
+    def send(self, delta, sample_rate):
+        """Send the counter to the backend.
+
+        :param float delta: The amount to increase the counter by.
+        :param float sample_rate: What rate this counter is sampled at. [0-1].
+
         """
         parts = [
             self.name + (":{:g}".format(delta).encode()),
             b"c",
         ]
 
-        if sample_rate and sample_rate != 1.0:
+        if sample_rate != 1.0:
             parts.append("@{:g}".format(sample_rate).encode())
 
         serialized = b"|".join(parts)
         self.transport.send(serialized)
 
+
+class BatchCounter(Counter):
+    """Counter implementation that batches multiple increment calls.
+
+    The overall change for the counter is tracked in the :term:`delta`
+    attribute. The result of multiple calls is flushed to the stats
+    backend in one call.  The :term:`sample_rate` of the last increment or
+    decrement is used.
+
+    Example usage::
+        counter = BatchCounter(transport, "counter_name")
+        counter.increment()
+        do_something_else()
+        counter.increment()
+        counter.send()
+
+    The above example results in one packet indicating an increase of two
+    should be applied to "counter_name".
+    """
+    def __init__(self, transport, name):
+        self.transport = transport
+        self.name = name
+        self.delta = 0
+        self._sample_rate = 1.0
+
+    def increment(self, delta=1, sample_rate=1.0):
+        """Increment the counter.
+
+        :param float delta: The amount to increase the counter by.
+        :param float sample_rate: What rate this counter is sampled at. [0-1].
+
+        """
+        if sample_rate != self._sample_rate:
```

Comment:
Is this actually safe? This sets the sample rate to the most recently seen one without scaling previous values. So if I sent `increment(1, sample_rate=0.1)` and `increment(1, sample_rate=1)` it'd just send 2 @ 100% when it should really indicate 11 @ 100%. 

Should we internally normalize everything to a sample rate of 1? (That might mess up the derived metrics statsd makes?) Could we have multiple batches for each seen sample rate? Just raise an exception if the sample rate doesn't match the first seen?

Reply:
Ooof, yeah.

> Should we internally normalize everything to a sample rate of 1?

This feels disingenuous to users in addition to your comment on the metrics.

> Just raise an exception if the sample rate doesn't match the first seen?

This feels brittle.

> Could we have multiple batches for each seen sample rate?

This seems to make the most sense. Multiple sample rates will mean more data sent to the stats backend but we save when we can *and* stay true to the backend API.

I think it's straight forward enough for ``BatchCounter`` to keep a map of sample rates to deltas and iterate through those in the ``send`` method.

``increment`` and ``send`` would look something like:

```python

def increment(self, delta=1, sample_rate=1.0):
    packet = self.packets.get(sample_rate)  # I think floats can be used as keys?
    if packet is None:
        self.packets[sample_rate] = delta
    else:
        self.packets[sample_rate] = self.packets[sample_rate] + delta

def send(self):
    for sample_rate, delta in self.packet.items()
        super(BatchCounter, self).send(delta, sample_rate)

```

Thoughts @spladug?

Reply:
TBH in all our metrics usage here we've almost never touched sample rate so I'm not sure how much effort it's worth putting into it which is why I was suggesting just throwing an error and if it becomes a blocker down the line we can make it more complex and robust.

re: keying off sample rate, I always get a little worried using floats in a context like that since they're so fickle. Maybe it'd be better to serialize them to string like they'll be on the wire before using as a key? Maybe overly paranoid.